linux/devdoc/socket_transport_linux.md

```diff
@@ -0,0 +1,284 @@
+```
+
+`socket_transport_connect` shall connect to a specified endpoint.
+
+If `socket_transport` is `NULL`, `socket_transport_connect` shall fail and return a non-zero value.
+
+If `hostname` is `NULL`, `socket_transport_connect` shall fail and return a non-zero value.
+
+If `port` is `0`, `socket_transport_connect` shall fail and return a non-zero value.
+
+If the transport type is not `SOCKET_CLIENT`, `socket_transport_connect` shall fail and return a non-zero value.
+
+`socket_transport_connect` shall call `sm_open_begin` to begin the open.
+
+If `sm_open_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_connect` shall fail and return a non-zero value.
+
+`socket_transport_connect` shall call `socket` with the params `AF_INET`, `SOCK_STREAM` and `0`.
+
+`socket_transport_connect` shall call `connect` to connect to the endpoint.
+
+`socket_transport_connect` shall set the socket to non-blocking.
+
+If successful `socket_transport_connect` shall call `sm_open_end` with `true`.
+
+If any failure is encountered, `socket_transport_connect` shall call `sm_open_end` with `false`, fail and return a non-zero value.
+
+### socket_transport_disconnect
+
+```c
+MOCKABLE_FUNCTION(, void, socket_transport_disconnect, SOCKET_TRANSPORT_HANDLE, socket_transport);
+```
+
+`socket_transport_disconnect` shall disconnect a connected socked from its endpoint.
+
+If `socket_transport` is `NULL`, `socket_transport_disconnect` shall fail and return.
+
+`socket_transport_disconnect` shall call `sm_close_begin` to begin the closing process.
+
+If `sm_close_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_disconnect` shall fail and return.
+
+`socket_transport_disconnect` shall call `close` to disconnect the connected socket.
+
+`socket_transport_disconnect` shall call `sm_close_end`.
+
+### socket_transport_send
+
+```c
+MOCKABLE_FUNCTION(, SOCKET_SEND_RESULT, socket_transport_send, SOCKET_TRANSPORT_HANDLE, socket_transport, SOCKET_BUFFER*, payload, uint32_t, buffer_count, uint32_t*, bytes_written, uint32_t, flags, void*, overlapped_data);
+```
+
+If `socket_transport` is `NULL`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+If `payload` is `NULL`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+If `buffer_count` is `0`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+`socket_transport_send` shall call `sm_exec_begin`.
+
+If `sm_exec_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+For each buffer count in payload `socket_transport_send` shall call `send` to send data with `flags`.
+
+If `send` returns a value less then 0, `socket_transport_send` shall stop sending and return `SOCKET_SEND_ERROR`.
+
+- If the errno is equal to ECONNRESET, `socket_transport_send` shall return `SOCKET_SEND_SHUTDOWN`.
```

Comment:
>errno [](http://example.com/codeflow?start=9&length=5)

IS this the real errno or some sort of GetLastError?

Reply:
Oh wait, we are in linux world here.

Reply:
The linux world is a different place sometimes.